pkg/limits/stream_metadata.go

```diff
@@ -103,76 +108,143 @@ func (s *streamMetadata) Usage(tenant string, fn UsageFunc) {
 		}
 	}
 }
-func (s *streamMetadata) Store(tenant string, partitionID int32, streamHash, recTotalSize uint64, recordTime, bucketStart, bucketCutOff int64) {
+
+func (s *streamMetadata) StoreIf(tenant string, streams map[int32][]Stream, maxActiveStreams uint64, cutoff, bucketStart, bucketCutOff int64) (map[Reason][]uint64, uint64) {
 	i := s.getStripeIdx(tenant)
 
 	s.locks[i].Lock()
 	defer s.locks[i].Unlock()
 
-	// Initialize tenant map if it doesn't exist
 	if _, ok := s.stripes[i][tenant]; !ok {
-		s.stripes[i][tenant] = make(map[int32][]Stream)
+		s.stripes[i][tenant] = make(map[int32]map[uint64]Stream)
 	}
 
-	// Initialize partition map if it doesn't exist
-	if s.stripes[i][tenant][partitionID] == nil {
-		s.stripes[i][tenant][partitionID] = make([]Stream, 0)
-	}
+	var (
+		exceedLimits  = make(map[Reason][]uint64)
```

Comment:
I think this can be a slice of `len(streams)`, as it doesn't look like we really use the result as a map in `ingest_limits.go`, instead we just iterate over it and append to another slice?

Reply:
Actually we are using the key (i.e. `Reason`) as to feed the results:
```go
        var results []*logproto.ExceedsLimitsResult
	for reason, streamHashes := range exceedLimits {
		for _, streamHash := range streamHashes {
			results = append(results, &logproto.ExceedsLimitsResult{
				StreamHash: streamHash,
				Reason:     uint32(reason),
			})
		}
	}
```

Also we will need the key as a differentiator when building the rate limits on a follow up PR.

Reply:
You misunderstood my question I think! 😄 But the last sentence I think explains why this needs to be a map instead of slice of structs.

Reply:
Yes counting rate limited streams was what I had in mind. That's what we did in the frontend previously.